### PR TITLE
WIP-Update users API

### PR DESCRIPTION
**Pull request by Fatema-Moaiyadi**

### CODE REVIEW

File: handler/user_handler.go [hunk 1]
```diff
@@ -93,3 +95,64 @@ func ListUsersHandler(userService service.UserService) http.HandlerFunc {
...
 	}
 }
+
+func UpdateUsersHandler(userService service.UserService) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+
+		w.Header().Set("Content-Type", "application/json")
+
+		params := mux.Vars(r)
+		id, _ := strconv.Atoi(params["id"])
```
Comment (vkhichar) [inline handler/user_handler.go]: return bad request from here if Atoi returns an error

File: service/user_service.go [hunk 2]
```diff
@@ -66,3 +68,17 @@ func (service *userService) CreateUser(ctx context.Context, user domain.User) (*
 	//create user service
 	return nil, nil
 }
+
+func (service *userService) UpdateUserService(ctx context.Context, id int, req contract.UpdateUserRequest) (*domain.User, error) {
```
Comment (vkhichar) [inline service/user_service.go]: rename the method to `UpdateUser`

File: repository/user_repository.go [hunk 3]
```diff
@@ -4,20 +4,26 @@ import (
...
 	selectAllUsers      = "SELECT id, name, email, password, is_admin, created_at, updated_at FROM users"
+	getUserByIDQuery    = "SELECT id, name, email, password, is_admin, created_at, updated_at FROM users WHERE id = $1"
+	updateUserColumns   = "UPDATE users SET name = $1, password = $2, updated_at = $3 WHERE id = $4"
 )
 
 type UserRepository interface {
 	FindUser(ctx context.Context, email string) (*domain.User, error)
 	CreateUser(ctx context.Context, user domain.User) (*domain.User, error)
 	ListUsers(ctx context.Context) ([]domain.User, error)
+	UpdateUserRepo(ctx context.Context, id int, req contract.UpdateUserRequest) (*domain.User, error)
```
Comment (vkhichar) [inline repository/user_repository.go]: rename this method to `UpdateUser`

File: repository/user_repository.go [hunk 4]
```diff
@@ -66,3 +72,37 @@ func (repo *userRepo) CreateUser(ctx context.Context, user domain.User) (*domain
...
+
+	if req.Password == nil {
+		req.Password = &tempUser.Password
+	}
+
+	tx := repo.db.MustBegin()
+
+	tx.MustExec(updateUserColumns, *req.Name, *req.Password, time.Now(), id)
+
+	err1 := tx.Commit()
```
Comment (vkhichar) [inline repository/user_repository.go]: better variable name for 'err1'

Comment (vkhichar) [inline repository/user_repository.go]: commit and rollback should be in defer function

**Review by vkhichar — CHANGES_REQUESTED** [review]

File: customerrors/errors.go [hunk 5]
```diff
@@ -4,8 +4,8 @@ import "errors"
 
 var (
 	ErrInvalidEmailPassword = errors.New("invalid email or password")
-
-	NoUsersExist = errors.New("No users exist at present")
-
-	NoAssetsExist = errors.New("No assets exist")
+	NoUsersExist            = errors.New("No users exist at present")
```
Comment (wsushmita) [inline customerrors/errors.go:7]: explain the purpose of two diff errors NoUsersExist & UserDoesNotExist

File: handler/user_handler.go [hunk 6]
```diff
@@ -93,3 +94,132 @@ func ListUsersHandler(userService service.UserService) http.HandlerFunc {
...
+		id, errInConversion := strconv.Atoi(params["id"])
+
+		if errInConversion != nil {
+			fmt.Printf("handler: Error while parameter conversion. Error: %s", errInConversion)
+			w.WriteHeader(http.StatusBadRequest)
+			responseBytes, err := json.Marshal(contract.ErrorResponse{Error: "Error while parameter conversion"})
+			w.Write(responseBytes)
+
+			if err != nil {
+				fmt.Printf("handler: Error while Marshal,%s", err)
```
Comment (wsushmita) [inline handler/user_handler.go:113]: in case of error we should return

File: handler/user_handler.go [hunk 7]
```diff
@@ -93,3 +94,132 @@ func ListUsersHandler(userService service.UserService) http.HandlerFunc {
...
+		err := json.NewDecoder(r.Body).Decode(&req)
+
+		if err != nil {
+			fmt.Printf("handler:Error while decoding request for update, %s", err)
+			w.WriteHeader(http.StatusBadRequest)
+			responseBytes, err := json.Marshal(contract.ErrorResponse{Error: "invalid request"})
+			w.Write(responseBytes)
+
+			if err != nil {
+				fmt.Printf("handler: Error while Marshal,%s", err)
```
Comment (wsushmita) [inline handler/user_handler.go:129]: same as above

File: handler/user_handler.go [hunk 8]
```diff
@@ -93,3 +94,132 @@ func ListUsersHandler(userService service.UserService) http.HandlerFunc {
...
+			}
+			return
+		}
+
+		user, err := userService.UpdateUser(r.Context(), id, req)
+
+		if err == customerrors.UserDoesNotExist {
```
Comment (wsushmita) [inline handler/user_handler.go:136]: err check should be in this order
if err != nil {
if err == customererr {
log error
return
}
log error 
return
}

pls make this change wherever neccessary

File: handler/user_handler.go [hunk 9]
```diff
@@ -93,3 +94,132 @@ func ListUsersHandler(userService service.UserService) http.HandlerFunc {
...
+				return
+			}
+			w.Write(responseBytes)
+			return
+
+		}
+
+		user, err := userService.DeleteUser(r.Context(), id)
+
+		if err == customerrors.NoUserExistForDelete {
```
Comment (wsushmita) [inline handler/user_handler.go:193]: use same custom err whenever we don't find any matching record for user

File: handler/user_handler.go [hunk 10]
```diff
@@ -93,3 +94,132 @@ func ListUsersHandler(userService service.UserService) http.HandlerFunc {
...
+			return
+		}
+		responsebytes, jsonErr := json.Marshal(user)
+		if jsonErr != nil {
+			fmt.Printf("handler: Error while converting user to json.Error: %s", jsonErr)
+			return
+		}
+		w.WriteHeader(http.StatusOK)
+		w.Write(responsebytes)
+	}
```
Comment (wsushmita) [inline handler/user_handler.go:224]: missing return?

File: repository/user_repo_test.go [hunk 11]
```diff
@@ -0,0 +1,141 @@
+package repository_test
+
+// import (
```
Comment (wsushmita) [inline repository/user_repo_test.go]: remove commented code

File: repository/user_repository.go [hunk 12]
```diff
@@ -4,20 +4,28 @@ import (
...
+	"github.com/vkhichar/asset-management/customerrors"
 	"github.com/vkhichar/asset-management/domain"
 )
 
 const (
 	getUserByEmailQuery = "SELECT id, name, email, password, is_admin FROM users WHERE email= $1"
 	selectAllUsers      = "SELECT id, name, email, password, is_admin, created_at, updated_at FROM users"
+	getUserByIDQuery    = "SELECT id, name, email, password, is_admin, created_at, updated_at FROM users WHERE id = $1"
+	updateUserColumns   = "UPDATE users SET name = $1, password = $2, updated_at = $3 WHERE id = $4"
+	deleteUserById      = "DELETE FROM users WHERE id=$1"
```
Comment (wsushmita) [inline repository/user_repository.go:20]: delete user should be soft delete here

File: repository/user_repository.go [hunk 13]
```diff
@@ -66,3 +74,53 @@ func (repo *userRepo) CreateUser(ctx context.Context, user domain.User) (*domain
...
+}
+
+func (repo *userRepo) DeleteUser(ctx context.Context, id int) (*domain.User, error) {
+	var user domain.User
+
+	err := repo.db.Get(&user, getUserByIDQuery, id)
+
+	if err == sql.ErrNoRows {
+		fmt.Printf("Repository: No users present")
+		return nil, nil
```
Comment (wsushmita) [inline repository/user_repository.go:114]: return err here

File: repository/user_repository.go [hunk 14]
```diff
@@ -66,3 +74,53 @@ func (repo *userRepo) CreateUser(ctx context.Context, user domain.User) (*domain
...
+	}
+	return &user, nil
+}
+
+func (repo *userRepo) DeleteUser(ctx context.Context, id int) (*domain.User, error) {
+	var user domain.User
+
+	err := repo.db.Get(&user, getUserByIDQuery, id)
+
+	if err == sql.ErrNoRows {
```
Comment (wsushmita) [inline repository/user_repository.go:112]: check if sequence mentioned in comments above

**Review by wsushmita — CHANGES_REQUESTED** [review]

File: customerrors/errors.go (hunk 5 above)
Comment (Fatema-Moaiyadi) [reply customerrors/errors.go:7]: I had made this from different branches. Will change code to keep only one now.

File: repository/user_repository.go (hunk 12 above)
Comment (Fatema-Moaiyadi) [reply repository/user_repository.go:20]: For user table we decided to permanently delete that user from table.

**Pull request merged**